### PR TITLE
Add learned rewrite preprocessing pass

### PR DESCRIPTION
**Pull request by ajreynol**
Adds the basic skeleton of the pass.

### CODE REVIEW

File: src/preprocessing/passes/learned_rewrite.cpp [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    case LearnedRewriteId::NONE: return "NONE";
+    default: return "?LearnedRewriteId?";
+  }
+}
+
+std::ostream& operator<<(std::ostream& out, LearnedRewriteId i)
+{
+  out << toString(i);
+  return out;
+}
```
Comment (yoni206) [inline src/preprocessing/passes/learned_rewrite.cpp:51]: Does this guideline about adding an empty namespace apply here?
https://github.com/cvc5/cvc5/wiki/Code-Style-Guidelines#unnamed-namespaces-and-static-variables

Do we really need these in the h file (in your future PRs on this)?

File: src/preprocessing/passes/learned_rewrite.cpp [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+  {
+    Trace("learned-rewrite-ll") << "No learned literals" << std::endl;
+    return PreprocessingPassResult::NO_CONFLICT;
+  }
+  else
+  {
+    Trace("learned-rewrite-ll") << "Learned literals:" << std::endl;
+    for (const Node& l : learnedLits)
+    {
+      // maybe for bound inference?
```
Comment (yoni206) [inline src/preprocessing/passes/learned_rewrite.cpp:77]:
```suggestion
      // maybe use bound inference?
```

File: src/preprocessing/passes/learned_rewrite.cpp [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+      Kind k = l.getKind();
+      if (k == EQUAL || k == GEQ)
+      {
+        binfer.add(l);
+      }
+      Trace("learned-rewrite-ll") << "- " << l << std::endl;
+    }
+    const std::map<Node, arith::Bounds>& bs = binfer.get();
+    // get the literals that were critical
```
Comment (yoni206) [inline src/preprocessing/passes/learned_rewrite.cpp:86]: Can u explain a little bit in a comment what does critical mean?

**Review by yoni206 — APPROVED** [review]
LGTM, modulo minor comments for your consideration.

File: src/preprocessing/passes/learned_rewrite.cpp [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+  {
+    Trace("learned-rewrite-ll") << "Learned literals:" << std::endl;
+    for (const Node& l : learnedLits)
+    {
+      // maybe for bound inference?
+      Kind k = l.getKind();
+      if (k == EQUAL || k == GEQ)
+      {
+        binfer.add(l);
+      }
```
Comment (nafur) [inline src/preprocessing/passes/learned_rewrite.cpp:82]: Why don't we use other kinds?

File: src/preprocessing/passes/learned_rewrite.cpp [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+      Trace("learned-rewrite-ll") << "- " << l << std::endl;
+    }
+    const std::map<Node, arith::Bounds>& bs = binfer.get();
+    // get the literals that were critical
+    for (const std::pair<const Node, arith::Bounds>& b : bs)
+    {
+      for (size_t i = 0; i < 2; i++)
+      {
+        Node origin = i == 0 ? b.second.lower_origin : b.second.upper_origin;
```
Comment (nafur) [inline src/preprocessing/passes/learned_rewrite.cpp:91]:
```suggestion
      for (const auto& origin: {b.second.lower_origin, b.second.upper_origin})
      {
```

File: src/preprocessing/passes/learned_rewrite.cpp [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+  }
+  // add the conjunction of learned literals back to assertions
+  if (!llrw.empty())
+  {
+    NodeManager* nm = NodeManager::currentNM();
+    std::vector<Node> llrvec(llrw.begin(), llrw.end());
+    Node llc = nm->mkAnd(llrvec);
+    Trace("learned-rewrite-assert")
+        << "Re-add rewritten learned conjunction: " << llc << std::endl;
+    assertionsToPreprocess->push_back(llc);
```
Comment (nafur) [inline src/preprocessing/passes/learned_rewrite.cpp:143]: Does this also add literals that have not changed (i.e. are already in the assertion list)?

File: src/preprocessing/passes/learned_rewrite.cpp (hunk 1 above)
Comment (ajreynol) [reply src/preprocessing/passes/learned_rewrite.cpp:51]: Yes, this is needed for statistics.

Comment (ajreynol) [reply src/preprocessing/passes/learned_rewrite.cpp:51]: As far as I know, the empty namespace suggestion does not apply.

File: src/preprocessing/passes/learned_rewrite.cpp (hunk 4 above)
Comment (ajreynol) [reply src/preprocessing/passes/learned_rewrite.cpp:82]: They are rewritten away, I'll add an assertion.

File: src/preprocessing/passes/learned_rewrite.cpp (hunk 6 above)
Comment (ajreynol) [reply src/preprocessing/passes/learned_rewrite.cpp:143]: Yes, potentially.

File: src/preprocessing/passes/learned_rewrite.cpp (hunk 4 above)
Comment (nafur) [reply src/preprocessing/passes/learned_rewrite.cpp:83]: Can there be negated ones?

Comment (ajreynol) [reply src/preprocessing/passes/learned_rewrite.cpp:83]: ah, good point, I'll fix this on my next PR

**Pull request merged**